Radon/iradon: rotation center re-definition & user definition support

Obi-Wan opened the pull request:
## Description

The current center definition in `radon`, `iradon`, and `iradon_sart` changes, depending on the image shape, and it is not coherent with the function `rotate`. The current definition is `image.shape // 2`, which always aligns with an integer pixel position. The function `rotate` and various tomography libraries use the definition `(image.shape - 1) / 2`, which always puts the rotation axis in the center of the image.

This PR changes the definition of the rotation center to `(image.shape - 1) / 2`, and it add support for specifying the center as an argument to those functions. This way, they users can either specify off-center rotation axes, or fall back to the previous behavior.

This closes #6425.

## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [x] Gallery example in `./doc/examples` (new features only)
<!-- - Benchmark in `./benchmarks`, if your changes aren't covered by an existing benchmark -->
- [x] Unit tests
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] Descriptive commit messages (see below)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

Obi-Wan committed:
Radon/iradon: define center as in rotate, and add support for user defined center

Signed-off-by: Nicola VIGANO <ben.vighy@gmail.com>

pep8speaks commented:
Hello @Obi-Wan! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`skimage/transform/radon_transform.py`](https://github.com/scikit-image/scikit-image/blob/f817d2a2949725af8530c59ffd29144118f53f70/skimage/transform/radon_transform.py):

> [Line 16:80](https://github.com/scikit-image/scikit-image/blob/f817d2a2949725af8530c59ffd29144118f53f70/skimage/transform/radon_transform.py#L16): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 234:80](https://github.com/scikit-image/scikit-image/blob/f817d2a2949725af8530c59ffd29144118f53f70/skimage/transform/radon_transform.py#L234): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 262:80](https://github.com/scikit-image/scikit-image/blob/f817d2a2949725af8530c59ffd29144118f53f70/skimage/transform/radon_transform.py#L262): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 443:80](https://github.com/scikit-image/scikit-image/blob/f817d2a2949725af8530c59ffd29144118f53f70/skimage/transform/radon_transform.py#L443): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)

* In the file [`skimage/transform/tests/test_radon_transform.py`](https://github.com/scikit-image/scikit-image/blob/f817d2a2949725af8530c59ffd29144118f53f70/skimage/transform/tests/test_radon_transform.py):

> [Line 242:54](https://github.com/scikit-image/scikit-image/blob/f817d2a2949725af8530c59ffd29144118f53f70/skimage/transform/tests/test_radon_transform.py#L242): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator

Obi-Wan commented:
Thank you @mkcor ! I will try to fix the conflicts by the end of the week / beginning of next week.

Obi-Wan commented:
@mkcor @lagru sorry for the delay. I ended up being much busier than expected. Moreover, the build system has completely changed since my initial submission. The PR is now ready for revision.

I have fixed the patch, and tried to make it as least invasive as possible.
Here are the main changes:
- Added the parameter `center=None` to `radon`, `iradon`, and `iradon_sart`
- Changed default center position from `(center // 2)` to `(center - 1) / 2` to be coherent with `rotate`
- Added tests for both default and non-default center position
- Fixed a bug in an unrelated test (`test_iradon_angles`)
- Changed docstring accordingly

All tests pass on my side. There is only one thing called "Labels" that doesn't pass, and I have no idea what it is.

Obi-Wan committed:
Fixed tests (circle to square)

Signed-off-by: Nicola VIGANO <ben.vighy@gmail.com>

Obi-Wan committed:
Fixed pre-commit

Signed-off-by: Nicola VIGANO <ben.vighy@gmail.com>

lagru commented:
Thanks for continued work on this! 

> Changed default center position from (center // 2) to (center - 1) / 2 to be coherent with rotate

This change might be a problem? This reads like the result might change in some cases between `main` and this branch even if the user didn't change anything? We have a policy in scikit-image not to do that unless we consider the previous output a bug. Or am I missing something?

Please feel welcome to request a review from me if you feel that it's ready. :)

mkcor commented:
> This change might be a problem? This reads like the result might change in some cases between `main` and this branch even if the user didn't change anything?

Right; I just started re-reviewing thinking that we were going for non-breaking changes https://github.com/scikit-image/scikit-image/issues/6425#issuecomment-1182272970 but, reading the entire discussion at #6425, I noticed that I eventually agreed https://github.com/scikit-image/scikit-image/issues/6425#issuecomment-1184323758 that this breaking change could be good *for skimage2*...

@Obi-Wan which means that: We cannot merge the PR into `main` as is; either we keep it on the backburner for skimage2 (which takes us towards the end of 2024), which was always option n. 1; or we turn this PR into a backwards-compatible version (which was option n. 2) and we review/approve/merge it quickly.

stefanv commented:
Why, in the past, did we not use the actual center of the image? This feel incorrect, but I haven't delved into the matter deeply.

Obi-Wan commented:
@mkcor @lagru ok, I will roll back the default center of rotation. The addition of the `center` parameter would make the API powerful enough to compensate/mitigate for the peculiar choice of the default value (when necessary).

I want this PR merged sooner than later, because if you like the idea, I would then make a new PR to add support for the projection shifts to both `radon` and `iradon`, as they are already supported in `iradon_sart`.

Finally, if you agree, we could add a disclaimer in the docstrings, which warns about a future change of the default.

stefanv commented:
@lagru Perhaps a 2.0 fix, to change the center?

mkcor commented:
> @lagru Perhaps a 2.0 fix, to change the center?

Yes, well, that was the conclusion reached at #6425.

mkcor commented:
> @mkcor @lagru ok, I will roll back the default center of rotation. The addition of the `center` parameter would make the API powerful enough to compensate/mitigate for the peculiar choice of the default value (when necessary).
> 
> I want this PR merged sooner than later

Ok, this is possible, since such an API change wouldn't break existing code nor change existing results.

Obi-Wan commented on skimage/transform/radon_transform.py:
```diff
@@ -34,12 +33,16 @@ def radon(image, theta=None, circle=True, *, preserve_range=False):
     radon_image : ndarray
         Radon transform (sinogram).  The tomography rotation axis will lie
-        at the pixel index ``radon_image.shape[0] // 2`` along the 0th
+        at the position ``(radon_image.shape[0] - 1) / 2`` along the 0th
```
You raise a good point. Changing the rotation axis to `(radon_image.shape[0] - 1) / 2`, that won't be true for even-sized images. However, it is really not a big deal...

Obi-Wan replied:
Well, not necessarily anymore. I will change accordingly, and refer to the `center` parameter.

Obi-Wan committed:
Apply suggestions from code review

This only affects docstrings. Code changes will follow

Co-authored-by: Marianne Corvellec <marianne.corvellec@ens-lyon.org>

Obi-Wan committed:
Reverted default center position, but parametrized with a function

Signed-off-by: Nicola VIGANO <ben.vighy@gmail.com>

Obi-Wan commented:
Ok, sorry for the long reaction time. I was distracted by some important personal matters.

@mkcor @lagru The requested revert of the default rotation center is now in place. I have done it through the addition of a small private function, for the following reasons:
- It will be easier to switch to the new definition at the time of skimage2 (it is just a boolean flag) - no need to find every place where its definition is implicitly used.
- The chosen center definition can be directly pulled by the unit tests.
- It could be reused in skimage2 to expose the legacy convention, in a practical manner (without requiring the users to do the math themselves).

The only small limitation from the above is the Cython code, which requires an equivalent function for the definition of the default center. I might be able to remove it, by leveraging the `projection_shifts`, but it might be too much for this PR. I will reserve it for the future PR dedicated to enabling the same support for `projection_shifts` in `radon` and `iradon` as it is exposed in `iradon_sart`.

mkcor commented on skimage/transform/tests/test_radon_transform.py:
```diff
@@ -66,47 +67,89 @@ def test_iradon_bias_circular_phantom():
 
 
-def check_radon_center(shape, circle, dtype, preserve_range):
+def check_radon_center(shape, circle, dtype, preserve_range, center_disp):
```
It's funny that, in this file, some test functions have names starting with `check_` (instead of the usual `test_`)...

Obi-Wan replied:
Yes, those `check_` functions are then called by `test_` functions, in a parametrized manner. Some of the `check_` functions are called by different `test_` functions. I suppose they used it to differentiate between the functions doing the job, and the ones selecting the actual configuration to test.
Do we need to maybe change the `check_` functions to something like `_test_`?

lagru replied:
No, if it works best to leave it, possibly for another PR if the need arises to clean this up.

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -13,16 +13,22 @@
+
+
+def radon(image, theta=None, circle=True, *, preserve_range=False, center=None):
     """
     Calculates the radon transform of an image given specified
```
```suggestion
    """Calculate the Radon transform of an image given the
```
Doesn't come from this PR, but we might as well fix this minor issue.

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -63,62 +75,75 @@ def radon(image, theta=None, circle=True, *, preserve_range=False):
     image = convert_to_float(image, preserve_range)
 
+    def_rot_center = _get_default_center(np.array(image.shape, dtype=np.float32))
+    if center is not None:
```
Please start with the default condition, `if center is None`.

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -63,62 +75,75 @@ def radon(image, theta=None, circle=True, *, preserve_range=False):
 
+    def_rot_center = _get_default_center(np.array(image.shape, dtype=np.float32))
+    if center is not None:
+        center = np.array(center, dtype=np.float32)
```
Why this dtype? I think you want to use `np.asarray` here (see above).

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -63,62 +75,75 @@ def radon(image, theta=None, circle=True, *, preserve_range=False):
+
     radon_image = np.zeros((padded_image.shape[0], len(theta)), dtype=image.dtype)
 
+    padded_center = _get_default_center(np.array(padded_image.shape, dtype=np.float32))
```
```suggestion
    padded_center = _get_default_center(np.asarray(padded_image.shape))
```

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -63,62 +75,75 @@ def radon(image, theta=None, circle=True, *, preserve_range=False):
+        img_shape = np.array(image.shape, dtype=int)
         coords = np.array(np.ogrid[: image.shape[0], : image.shape[1]], dtype=object)
-        dist = ((coords - img_shape // 2) ** 2).sum(0)
+        dist = ((coords - (def_rot_center + rel_rot_center)) ** 2).sum(axis=0)
```
The innermost parenthesis is basically `center` -- I believe it would be much more readable to have `center` instead of `(def_rot_center + rel_rot_center)`. You just have to define `center` in the (default) case where `center is None`.

mkcor reviewed (commented):
No worries, @Obi-Wan!

Obi-Wan commented on skimage/transform/radon_transform.py:
```diff
@@ -13,16 +13,22 @@
 
 
-def radon(image, theta=None, circle=True, *, preserve_range=False):
+def _get_default_center(image_shape, legacy: bool = True):
```
Yes, when adding type hints, the convention changes. This can be seen for instance here: https://docs.python.org/3/library/typing.html#typing.Optional
Moreover, `black` systematically adds the spaces in there.

lagru replied:
```suggestion
def _get_default_center(image_shape, legacy=True):
```
We are currently not using type hints. We are planning to but those wouldn't be inline. So this lonely type hint should be removed. :smile:

mkcor replied:
@Obi-Wan good to know, thanks! Please ignore my suggestion and, for this PR, include @lagru's.

lagru commented on skimage/transform/radon_transform.py:
```diff
@@ -34,13 +40,19 @@ def radon(image, theta=None, circle=True, *, preserve_range=False):
+        Coordinates of the rotation axis. If None, the rotation axis will be
+        located in the pixel with coordinates:
+        ``(image.shape[0] // 2, image.shape[1] // 2)``
+        This will change in a future release, and it will be moved to:
```
Hmm, I wouldn't. If we change the plan around skimage2 again (hopefully not) we would probably just forget this. :sweat_smile:

mkcor commented on skimage/transform/_radon_transform.pyx:
```diff
@@ -33,16 +39,17 @@ cdef bilinear_ray_sum(np_floats[:, :] image, np_floats theta,
     s0 = sqrt(radius * radius - t * t) if radius*radius >= t*t else 0.
     cdef Py_ssize_t Ns = 2 * (<Py_ssize_t>ceil(2 * s0))  # number of steps
                                                          # along the ray
+
```
Was the addition of this blank line intentional?

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -224,13 +247,20 @@ def iradon(
+        affects where it is supposed to be in the radon image. By default, the
+        rotation axis will be located in the reconstruction at the position:
+        ``(reconstructed.shape[0] // 2, reconstructed.shape[1] // 2)``
+        This will change in a future release, and it will be moved to:
```
```suggestion
        In a future release, this will be changed to:
```

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -34,13 +40,19 @@ def radon(image, theta=None, circle=True, *, preserve_range=False):
-        at the pixel index ``radon_image.shape[0] // 2`` along the 0th
-        dimension of ``radon_image``.
+        by default at the position ``radon_image.shape[0] // 2`` along the 0th
+        dimension of ``radon_image``. This will change in a future release.
```
```suggestion
        axis of ``radon_image``. This will change in a future release.
```
"In NumPy dimensions are called axes." ([ref](https://numpy.org/devdocs/user/quickstart.html#the-basics)).

mkcor replied:
But I admit that, in this particular case, this may cause confusion with the *rotation axis* we keep talking about...

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -224,13 +247,20 @@ def iradon(
+        This will change in a future release, and it will be moved to:
+        ``((reconstructed.shape[0] - 1) / 2, (reconstructed.shape[1] - 1) / 2)``
+        Accordingly, it will be located at the position
+        ``radon_image.shape[0] // 2`` along the 0th dimension of ``radon_image``.
```
```suggestion
        ``radon_image.shape[0] // 2`` along the 0th axis of ``radon_image``.
```

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -399,9 +447,9 @@ def iradon_sart(
         the image corresponds to a projection along a different angle. The
-        tomography rotation axis should lie at the pixel index
+        tomography rotation axis will lie by default at the position
         ``radon_image.shape[0] // 2`` along the 0th dimension of
```
```suggestion
        ``radon_image.shape[0] // 2`` along the 0th axis of
```

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -224,13 +247,20 @@ def iradon(
         image is converted according to the conventions of `img_as_float`.
         Also see https://scikit-image.org/docs/dev/user_guide/data_types.html
+    center : array_like, optional
+        It specifies the rotation center of the reconstructed image. This also
```
```suggestion
        Position of the rotation center in the reconstructed image. Changing it
```

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -224,13 +247,20 @@ def iradon(
         Also see https://scikit-image.org/docs/dev/user_guide/data_types.html
+    center : array_like, optional
+        It specifies the rotation center of the reconstructed image. This also
+        affects where it is supposed to be in the radon image. By default, the
```
```suggestion
        will affect its position in the radon image as well. By default, the
```

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -224,13 +247,20 @@ def iradon(
+    center : array_like, optional
+        It specifies the rotation center of the reconstructed image. This also
+        affects where it is supposed to be in the radon image. By default, the
+        rotation axis will be located in the reconstruction at the position:
```
```suggestion
        the position of rotation axis in the reconstruction is located at:
```

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -224,13 +247,20 @@ def iradon(
+        ``(reconstructed.shape[0] // 2, reconstructed.shape[1] // 2)``
+        This will change in a future release, and it will be moved to:
+        ``((reconstructed.shape[0] - 1) / 2, (reconstructed.shape[1] - 1) / 2)``
+        Accordingly, it will be located at the position
```
I think that it's not clear whether 'Accordingly' refers to the current default or the future default...

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -257,8 +287,15 @@ def iradon(
         theta = np.linspace(0, 180, radon_image.shape[1], endpoint=False)
 
+    img_shape, radon_angles = radon_image.shape
+    if center is not None:
```
Same comment as above (regarding the same conditional).

mkcor commented on skimage/transform/radon_transform.py:
```diff
@@ -422,13 +470,20 @@ def iradon_sart(
         data type is not float, input is cast to double, otherwise
         dtype is set to input data type.
+    center : array_like, optional
+        It specifies the rotation center of the reconstructed image. This also
```
Same suggestions as above.

mkcor reviewed (commented):
We're getting close! Thank you for your patience, @Obi-Wan. Please let us know if you need help addressing the review comments.

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
